### PR TITLE
Support storing Interval types in tables

### PR DESCRIPTION
**Pull request by larryluogit**
### Changelog category (leave one):

- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

- Support storing `Interval` data types in tables.

Interval types are integers, but with richer metadata (unit). Datetime arithmetic is the most common use cases for Interval. However, Interval can be use in another use case as well. One application can generate a series of data and store it in a table as any Interval type. Another application can inspect the temporary table and understand what unit those integers are.

### CODE REVIEW

**Review by vdimir — CHANGES_REQUESTED** [review]
Thank you for your PR!
Could you please add more motivation for this change, show more examples and use-cases, consider possible pitfalls?

> One application can generate a series of data and store it in a table as any Interval type. Another application can inspect the temporary table and understand what unit those integers are.

You can always store just numbers in the table and use them as intervals. In such a case, all intervals should be of the same kind, but the user can always convert them to suitable values before inserting. 

```sql
SELECT now() - INTERVAL number DAY FROM numbers(10);


┌─minus(now(), toIntervalDay(number))─┐
│                 2023-04-24 10:58:10 │
│                 2023-04-23 10:58:10 │
│                 2023-04-22 10:58:10 │
│                 2023-04-21 10:58:10 │
│                 2023-04-20 10:58:10 │
│                 2023-04-19 10:58:10 │
│                 2023-04-18 10:58:10 │
│                 2023-04-17 10:58:10 │
│                 2023-04-16 10:58:10 │
│                 2023-04-15 10:58:10 │
└─────────────────────────────────────┘
```


I'm afraid that storing the interval type in the table may be confusing, as the INTERVAL type doesn't have proper formatting. Additionally, type names for different intervals are different: toTypeName(INTERVAL number DAY) -> IntervalDay and toTypeName(INTERVAL number WEEK) -> IntervalWeek."

**Comment by larryluogit** [comment]
> Thank you for your PR! Could you please add more motivation for this change, show more examples and use-cases, consider possible pitfalls?
> 
> > One application can generate a series of data and store it in a table as any Interval type. Another application can inspect the temporary table and understand what unit those integers are.
> 
> You can always store just numbers in the table and use them as intervals. In such a case, all intervals should be of the same kind, but the user can always convert them to suitable values before inserting.
> 
> ```sql
> SELECT now() - INTERVAL number DAY FROM numbers(10);
> 
> 
> ┌─minus(now(), toIntervalDay(number))─┐
> │                 2023-04-24 10:58:10 │
> │                 2023-04-23 10:58:10 │
> │                 2023-04-22 10:58:10 │
> │                 2023-04-21 10:58:10 │
> │                 2023-04-20 10:58:10 │
> │                 2023-04-19 10:58:10 │
> │                 2023-04-18 10:58:10 │
> │                 2023-04-17 10:58:10 │
> │                 2023-04-16 10:58:10 │
> │                 2023-04-15 10:58:10 │
> └─────────────────────────────────────┘
> ```
> 
> I'm afraid that storing the interval type in the table may be confusing, as the INTERVAL type doesn't have proper formatting. Additionally, type names for different intervals are different: toTypeName(INTERVAL number DAY) -> IntervalDay and toTypeName(INTERVAL number WEEK) -> IntervalWeek."

Thanks @vdimir 
The motivation of this change is to simplify metadata communication between data generator and consumer. One application can generate a series of data and store it in a able as any Interval type. Another application can use the interval data without explicit casting. If pure integer is used, then the consumer of the data must have knowledge of what those integer mean, and do an explicit casting before using it. The two applications has to setup another way to communicate the extra metadata (unit of interval).
Here is a slack conversation that can provide more context.
https://clickhousedb.slack.com/archives/CUDSPUJ68/p1681913041054419

**Comment by larryluogit** [comment]
> 



> Also, in ClickHouse combination of intervals is tuple of intervals, but for other databases, it's just simple Interval data type, which can store different granularity underneath.
> 
> ```
> SELECT toTypeName(toIntervalDay(1) + toIntervalSecond(1))
> 
> Query id: 7d53c8cd-fbda-487a-8f65-cc0823f00070
> 
> ┌─toTypeName(plus(toIntervalDay(1), toIntervalSecond(1)))─┐
> │ Tuple(IntervalDay, IntervalSecond)                      │
> └─────────────────────────────────────────────────────────┘
> ```

Thanks for pointing out combination case. This is another example saving interval allows consumer to use the complex data directly without casting.
```
create table t engine=Memory as SELECT toIntervalDay(1) + toIntervalSecond(1) as v;

with toDateTime64('2023-01-01 00:00:00.000000001', 9, 'US/Eastern') as c select c+v.1+v.2 from t;

┌─plus(plus(c, tupleElement(v, 1)), tupleElement(v, 2))─┐
│                         2023-01-02 00:00:01.000000001 │
└───────────────────────────────────────────────────────┘
```

**Comment by larryluogit** [comment]
@vdimir Is there any other information I should add to this PR?

**Comment by robot-ch-test-poll3** [comment]
This is an automated comment for commit e18395b265927d9099ca7e51a6215e4c97be97a8 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49085/e18395b265927d9099ca7e51a6215e4c97be97a8/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by vdimir** [comment]
@larryluogit sorry for the late reply.

I've taken another look, and while I understand the motivation behind it, in my opinion, it is not entirely strong.
We still should define the interval type that will be stored in the table before creating it. For instance, if we choose minutes as the interval type, we cannot insert seconds and must convert them to minutes first. Therefore, there isn't much convenience in favor of using interval types over raw numbers. The only motivation I see is that it prevents misuse in some cases, as it forces users to explicitly convert their values.

```sql

:) create table saved_intervals Engine = Memory as SELECT toIntervalMinute(number+1) as v1 from numbers(10);

Ok.

0 rows in set. Elapsed: 0.006 sec. 

:) INSERT INTO saved_intervals SELECT toIntervalSecond(number+1) as v1 from numbers(10);

0 rows in set. Elapsed: 0.037 sec. 

Received exception from server (version 23.5.1):
Code: 43. DB::Exception: Received from localhost:9000. DB::Exception: Illegal type IntervalSecond of argument of function toIntervalMinute: while converting source column v1 to destination column v1. (ILLEGAL_TYPE_OF_ARGUMENT)

:) DESC saved_intervals;

DESCRIBE TABLE saved_intervals

┌─name─┬─type───────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ v1   │ IntervalMinute │              │                    │         │                  │                │
└──────┴────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘
```

File: tests/queries/0_stateless/02724_persist_interval_type.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
+DROP TABLE IF EXISTS saved_intervals_tmp;
+DROP TABLE IF EXISTS saved_intervals_mgt;
+create table saved_intervals_tmp Engine=Memory as SELECT number as EventID, toIntervalSecond(number+1) as v1, toIntervalHour(number+2) as v2, toIntervalNanosecond(number+3) as v3 from numbers(2);
```
Comment (vdimir) [inline tests/queries/0_stateless/02724_persist_interval_type.sql:4]: Let's also add CREATE table with explicit type definition like
```sql
CREATE table t1 (v1 IntervalMinute) ENGINE = Memory;
```

**Review by vdimir — CHANGES_REQUESTED** [review]
Well, I don't think they're any significant problems with it, we can merge it. Let's just reformat test and add more cases, e.g. with explicit table creating and also some corner cases like inserting `toIntervalMinute` to table with seconds .

**Comment by larryluogit** [comment]
> Well, I don't think they're any significant problems with it, we can merge it. Let's just reformat test and add more cases, e.g. with explicit table creating and also some corner cases like inserting `toIntervalMinute` to table with seconds .

@vdimir Thanks for reviewing the PR. I added new tests as suggested.
The actual use case is simpler. A system needs to hand over user selected data as read-only table to a different user. This limitation makes such system unnecessarily complicated. 
Besides the use case, it also makes sense to storing Interval types from consistency point of view. Normally, user expects to be able to save the result of a valid SELECT statement. Because of this limitation, user can not save the result if any of the columns in the result is Interval type. I’m not sure if any other databases have limitations in saving SELECT statement results.  I think removing this limitation makes the behaviour consistent in ClickHouse.

**Review by vdimir — APPROVED** [review]

File: tests/queries/0_stateless/02724_persist_interval_type.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+DROP TABLE IF EXISTS saved_intervals_mgt;
+create table saved_intervals_mgt Engine=MergeTree() ORDER BY EventID as SELECT number as EventID, toIntervalSecond(number+1) as v1, toIntervalHour(number+2) as v2, toIntervalNanosecond(number+3) as v3 from numbers(2);
+with toDateTime64('2023-01-01 00:00:00.000000001', 9, 'US/Eastern') as c select c+v1 as c_v1, c+v2 as c_v2, c+v3 as c_v3, date_diff(second, c, c_v1), date_diff(hour, c, c_v2), date_diff(second, c, c_v3) from saved_intervals_mgt;
+DROP TABLE IF EXISTS saved_intervals_mgt;
+
+DROP TABLE IF EXISTS t1;
+CREATE table t1 (v1 IntervalMinute) ENGINE = Memory;
+INSERT INTO t1 with toDateTime64('2023-01-01 00:00:00.000000001', 9, 'US/Eastern') as c SELECT EXTRACT(MINUTE FROM c+toIntervalSecond(number * 60)) from numbers(2);
+select * from t1;
+DROP TABLE IF EXISTS t1;
```
Comment (vdimir) [inline tests/queries/0_stateless/02724_persist_interval_type.sql:15]:
```suggestion
DROP TABLE IF EXISTS t1;

```

**Comment by vdimir** [comment]
> Stateless tests (release) — fail: 1, passed: 5282, skipped: 4 [Details](https://s3.amazonaws.com/clickhouse-test-reports/49085/4f3b7aded4fe33dc2c73b9af9cac5f9e877bc4b1/stateless_tests__release_.html)

Flaky `00002_log_and_exception_messages_formatting`, will skip mergable check and merge

**Comment by larryluogit** [comment]
@vdimir Can this PR be merged?

**Comment by vdimir** [comment]
@larryluogit yes, could you please rebase the branch (using the button in github web interface, for example)?

`Stateless tests (release) — fail: 1, passed: 5282, skipped: 4` failed because of flaky tests in master, but merge is unavailable because of this

**Pull request merged**